File: probs/1482.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int minDays(vector<int>& bloomDay, int m, int k) {
        int l = 0;
        int r = 0;
        for (int bloom: bloomDay) {
            r = max(r, bloom);
        }

        int best = -1;

        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (possible(bloomDay, m, k, mid)) {
                best = mid;
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }

        return best;
    }

private:
    bool possible(vector<int>& bloomDay, int m, int k, int day) {
        int now = 0;
        int count = 0;

        for (int i = 0; i < bloomDay.size(); i++) {
            if (bloomDay[i] <= day) {
                count++;
            } else {
                count = 0;
            }

            if (count == k) {
                now++;
                count = 0;
            }
        }

        return now >= m;
    }
};

File: probs/1482.rs
use std::cmp;

impl Solution {
    pub fn min_days(bloom_day: Vec<i32>, m: i32, k: i32) -> i32 {
        if bloom_day.len() < ((m * k) as usize) {
            return -1;
        }

        let mut lo = 0;
        let mut hi = 0;

        for day in bloom_day.iter() {
            hi = cmp::max(hi, *day);
        }

        let mut best = -1;

        while (lo <= hi) {
            let mid = lo + (hi - lo) / 2;
            if Self::possible(&bloom_day, m, k, mid) {
                best = mid;
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        }

        best
    }

    fn possible(bloom_day: &Vec<i32>, m: i32, k: i32, day: i32) -> bool {
        let mut bouquet = 0;
        let mut flowers = 0;

        for bloom in bloom_day.iter() {
            if *bloom <= day {
                flowers += 1;
            } else {
                flowers = 0;
            }

            if flowers == k {
                bouquet += 1;
                flowers = 0;
            }
        }

        bouquet >= m
    }
}
